containers.go

```diff
@@ -215,18 +215,41 @@ func (r *containerStore) stopWriting() {
 // If this succeeds, the caller MUST call stopReading().
 func (r *containerStore) startReading() error {
 	r.lockfile.RLock()
-	succeeded := false
+	unlockFn := r.lockfile.Unlock // A function to call to clean up, or nil
 	defer func() {
-		if !succeeded {
-			r.lockfile.Unlock()
+		if unlockFn != nil {
+			unlockFn()
 		}
 	}()
 
-	if err := r.reloadIfChanged(false); err != nil {
-		return err
+	if tryLockedForWriting, err := r.reloadIfChanged(false); err != nil {
+		if !tryLockedForWriting {
+			return err
+		}
+		unlockFn()
+		unlockFn = nil
+
+		r.lockfile.Lock()
```

Comment:
It looks like an upgradable lock (when the reader can become a writer but atomically). I didn't find a popular library for Go, but I have experience with [boost](https://www.boost.org/doc/libs/1_71_0/doc/html/interprocess/synchronization_mechanisms.html#interprocess.synchronization_mechanisms.sharable_upgradable_mutexes.sharable_upgradable_mutexes_operations.upgradable_mutexes_operations_upgradable) (just for example of upgradable mutex). Maybe it's good idea to use this approach here? I will allow avoiding double call of reloadIfChanged.

Reply:
Yes, a lock type like that can be built.

A limitation here is that we need an _inter-process_ lock, a lock that works across user namespaces without users having to set up anything (and if at all possible one that is interoperable with a previous version of c/storage). That basically limits us to the two kernel-provided file locking mechanisms, and other concerns direct us to the `fcntl` mechanism.

Now, `fcntl` locks do, it seems, support lock type change; the formal specification seems to be a bit ambiguous about whether it is atomic or whether it internally releases the lock.

---

Purely in principle:

For container and image stores, this is AFAICT a “shouldn’t happen” code path; in that case, performance really isn’t a concern at all, and it definitely doesn’t warrant adding a dependency or using a non-trivial locking implementation.

The layer store, with its recovery of incomplete layers, which _can_ happen in production for unavoidable reasons, an upgradeable would be useful because it would mean that we don’t have to loop, and don’t have to decide how much to loop. But, in that case, all locking costs are dwarfed by costs of actual cleanup filesystem operations anyway…

---

So, overall, I rather favor the most simple approach possible. Not that the loops in `layerStore.startReadingWithReload` are _that_ simple, but still, better than custom lock implementations. I guess?

Reply:
Thank you for the explanation. I agree, and I did not take into account that upgradable requires an inter-process lock.